File: nginx/log/log_rotate.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"
	"log"
	"strings"
)

// 获取当前日期，格式化为 YYYY-MM-DD
func getCurrentDate() string {
	return time.Now().Format("2006-01-02")
}

// 执行系统命令并检查返回值
func executeCommand(command string, args ...string) error {
	cmd := exec.Command(command, args...)
	err := cmd.Run()
	if err != nil {
		return fmt.Errorf("failed to execute command: %v", err)
	}
	return nil
}

// 删除超过7天的日志文件
func deleteOldLogs(logDir string) error {
	now := time.Now()

	err := filepath.Walk(logDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		// 只处理 .log.gz 文件
		if strings.HasSuffix(info.Name(), ".log.gz") {
			// 计算文件的修改时间与当前时间的差距
			if now.Sub(info.ModTime()).Hours() > 7*24*60*60 { // 超过7天
				err := os.Remove(path)
				if err != nil {
					return fmt.Errorf("failed to delete file %s: %v", path, err)
				}
				fmt.Printf("Deleted old log: %s\n", path)
			}
		}
		return nil
	})

	return err
}

func main() {
	logDir := "/data/docker_data/web/nginx/log"
	logDate := getCurrentDate()

	// 重命名日志文件
	accessLogOld := filepath.Join(logDir, "access.log")
	errorLogOld := filepath.Join(logDir, "error.log")
	accessLogNew := filepath.Join(logDir, fmt.Sprintf("access_%s.log", logDate))
	errorLogNew := filepath.Join(logDir, fmt.Sprintf("error_%s.log", logDate))

	err := os.Rename(accessLogOld, accessLogNew)
	if err != nil {
		log.Fatalf("Failed to rename access log: %v\n", err)
	}

	err = os.Rename(errorLogOld, errorLogNew)
	if err != nil {
		log.Fatalf("Failed to rename error log: %v\n", err)
	}

	// 检查 Nginx 容器是否存在
	_, err = exec.Command("docker", "inspect", "nginx").Output()
	if err == nil {
		// 向Nginx发送信号，重新打开日志文件
		err := executeCommand("docker", "exec", "nginx", "nginx", "-s", "reopen")
		if err != nil {
			log.Fatalf("Failed to reopen Nginx logs: %v\n", err)
		}
	} else {
		fmt.Println("Nginx container not found. Skipping nginx -s reopen.")
	}

	// 压缩旧日志
	err = executeCommand("gzip", accessLogNew)
	if err != nil {
		log.Fatalf("Failed to gzip access log: %v\n", err)
	}

	err = executeCommand("gzip", errorLogNew)
	if err != nil {
		log.Fatalf("Failed to gzip error log: %v\n", err)
	}

	// 删除7天前的日志文件
	err = deleteOldLogs(logDir)
	if err != nil {
		log.Fatalf("Failed to delete old logs: %v\n", err)
	}

	fmt.Println("Log rotation and cleanup completed successfully.")
}

File: nginx/log/log_rotate.pl
#!/usr/bin/env perl

use strict;
use warnings;
use File::Basename;
use POSIX qw(strftime);

# 设置日志目录和日期
my $log_dir = "/data/docker_data/web/nginx/log";
my $log_date = strftime("%Y-%m-%d", localtime);

# 切割日志
rename("$log_dir/access.log", "$log_dir/access_$log_date.log") or die "无法移动 access.log: $!\n";
rename("$log_dir/error.log", "$log_dir/error_$log_date.log") or die "无法移动 error.log: $!\n";

# 检查Nginx是否存在
my $nginx_running = `docker inspect "nginx" > /dev/null 2>&1; echo $?`;
chomp($nginx_running);

if ($nginx_running == 0) {
    # 向Nginx发送信号，重新打开日志文件
    system("docker exec nginx nginx -s reopen") == 0 or die "发送信号失败: $!\n";
} else {
    exit 0;
}

# 压缩旧日志
system("gzip", "$log_dir/access_$log_date.log") == 0 or die "压缩access.log失败: $!\n";
system("gzip", "$log_dir/error_$log_date.log") == 0 or die "压缩error.log失败: $!\n";

# 删除7天前的日志
opendir(my $dh, $log_dir) or die "无法打开目录 $log_dir: $!\n";
while (my $file = readdir($dh)) {
    if ($file =~ /\.log\.gz$/) {
        my $file_path = "$log_dir/$file";
        my $mtime = (stat($file_path))[9];  # 获取文件修改时间
        my $current_time = time;
        my $age_in_days = ($current_time - $mtime) / (60 * 60 * 24);

        if ($age_in_days > 7) {
            unlink($file_path) or warn "删除文件失败 $file_path: $!\n";
        }
    }
}
closedir($dh);